tests/framework/e2e/downgrade.go

```diff
@@ -162,15 +163,34 @@ func DowngradeUpgradeMembersByID(t *testing.T, lg *zap.Logger, clus *EtcdProcess
 		}
 	}
 
-	clusterVersion := targetVersion.String()
-	if !isDowngrade && len(membersToChange) != len(clus.Procs) {
-		clusterVersion = currentVersion.String()
-	}
+	t.Log("Waiting health interval to make sure the leader propagates version to new processes")
+	time.Sleep(etcdserver.HealthInterval)
+
 	lg.Info("Validating versions")
+	clusterVersion := targetVersion
+	if !isDowngrade {
+		// if there is any member using old version binary, the clusterVersion should be old version.
+		for _, member := range clus.Procs {
+			mVersion, err := GetVersionFromBinary(member.Config().ExecPath)
+			if err != nil {
+				return fmt.Errorf("failed to get version from member %s binary: %w", member.Config().Name, err)
+			}
+
+			if mVersion.LessThan(*clusterVersion) {
+				clusterVersion = mVersion
+				break
+			}
+		}
+
+		if downgradeEnabled {
+			clusterVersion = currentVersion
+		}
+	}
```

Comment:
Please see my simplification in https://github.com/etcd-io/etcd/pull/19461#issuecomment-2676095912